autoPyTorch/api/base_task.py

```diff
@@ -684,6 +685,269 @@ def _do_traditional_prediction(self, time_left: int, func_eval_time_limit_secs:
+            time_for_traditional = int(
+                self._total_walltime_limit - elapsed_time - self._func_eval_time_limit_secs
+            )
+            self._do_traditional_prediction(time_left=time_for_traditional)
+            self._stopwatch.stop_task(traditional_task_name)
+
+    def _run_ensemble(self, dataset: BaseDataset, optimize_metric: str,
+                      precision: int) -> Optional[EnsembleBuilderManager]:
+
+        assert self._logger is not None
+        assert self._metric is not None
+
+        elapsed_time = self._stopwatch.wall_elapsed(self.dataset_name)
+        time_left_for_ensembles = max(0, self._total_walltime_limit - elapsed_time)
+        proc_ensemble = None
+        if time_left_for_ensembles <= 0 and self.ensemble_size > 0:
+            raise ValueError("Could not run ensemble builder because there "
+                             "is no time left. Try increasing the value "
+                             "of total_walltime_limit.")
+        elif self.ensemble_size <= 0:
+            self._logger.info("Could not run ensemble builder as ensemble size is non-positive.")
+        else:
+            self._logger.info("Run ensemble")
+            ensemble_task_name = 'ensemble'
+            self._stopwatch.start_task(ensemble_task_name)
+            proc_ensemble = EnsembleBuilderManager(
+                start_time=time.time(),
+                time_left_for_ensembles=time_left_for_ensembles,
+                backend=copy.deepcopy(self._backend),
+                dataset_name=dataset.dataset_name,
+                output_type=STRING_TO_OUTPUT_TYPES[dataset.output_type],
+                task_type=STRING_TO_TASK_TYPES[self.task_type],
+                metrics=[self._metric],
+                opt_metric=optimize_metric,
+                ensemble_size=self.ensemble_size,
+                ensemble_nbest=self.ensemble_nbest,
+                max_models_on_disc=self.max_models_on_disc,
+                seed=self.seed,
+                max_iterations=None,
+                read_at_most=sys.maxsize,
+                ensemble_memory_limit=self._memory_limit,
+                random_state=self.seed,
+                precision=precision,
+                logger_port=self._logger_port
+            )
+            self._stopwatch.stop_task(ensemble_task_name)
+
+        return proc_ensemble
+
+    def _get_budget_config(self, budget_type: Optional[str] = None,
+                           budget: Optional[float] = None) -> Dict[str, Union[float, str]]:
+
+        budget_config: Dict[str, Union[float, str]] = {}
+        if budget_type is not None and budget is not None:
+            budget_config['budget_type'] = budget_type
+            budget_config[budget_type] = budget
+        elif budget_type is not None or budget is not None:
+            raise ValueError("budget type was not specified in budget_config")
+
+        return budget_config
+
+    def _start_smac(self, proc_smac: AutoMLSMBO) -> None:
```

Comment:
I think its fine if we have this function inside _run_smac. Having two functions makes it a bit confusing.